rfc/0000-configuration-and-shared-files.md

```diff
@@ -0,0 +1,264 @@
+
+It is up to the discretion of the language implementation to define the meaning of these variables. The implementation or tool may also deviate from this standard if deemed necessary or in order to follow language-specific conventions.
+
+The Pact Broker provides a good example of this guideline, where it uses `PACT_BROKER_*` to allow for many aspects of the broker to be configured. See the [Pact Broker Configuration docs](https://github.com/pact-foundation/pact_broker/blob/master/docs/CONFIGURATION.md) for more information of this use case.
+
+### Configuration
+
+This RFC proposes the use of [TOML](https://toml.io/en/) configuration files for storing configuration data. TOML is chosen due to its simplicity and readability. Other formats such as JSON or YAML could have been chosen, but they have the following drawbacks:
+
+- JSON is less human-friendly than TOML and does not support comments which can be useful for documenting configuration files.
+- YAML is more complex than TOML, and can be harder to parse correctly. It is also prone to errors due to the way it handles indentation and somewhat unconventional types (e.g., sexagesimal numbers where `1:2:3` equals `3723`, octal numbers where `0123` equals `83`, and automatic conversion of boolean-like values such as `no`, `on`, `y` into booleans).
+
+Configuration files complement the use of environment variables. They can allow for more complex configuration data to be stored, and can be used to store default values shared across the Pact ecosystem without requiring the end user to set environment variables all the time.
+
+This RFC also proposes to have both local (or per-project) configuration files, and user configuration files. This allows a user of Pact to configure their environment in a way that is consistent across all projects, as well as allowing for project-specific configuration.
+
+For example:
+
+- A user may wish to set the their default Pact Broker to `https://pact-broker.local.company.com` in their user configuration file so that all projects use this value by default.
+- A specific project might still be in development, and use a local Pact Broker at `http://localhost:9292` for testing. This could be set in the project's local configuration file.
+- The CI/CD pipeline still configures the Pact Broker using environment variables, as it is not practical to store configuration files in the pipeline.
+
+#### Precedence
+
+The following order of precedence _must_ be followed when reading configuration data:
+
+1. Environment variables
+2. Configuration files, in the following order:
+   1. If an explicit configuration file is set, only that file should be read.
+   2. If no explicit configuration file is set, the following files should be read in order:
+      1. Local configuration file
+      2. User configuration file
+3. Default values
+
+The explicit configuration file may either be set by the `PACT_CONFIG_FILE` environment variable, or by a command line option.
```

Comment:
Where does a CLI provided option rank in the list, if it differs from an env var?

Reply:
The CLI argument takes the priority, as it is explicitly set by the end user for a specific invocation and therefore is the _most_ specific. I'll amend the RFC to make this clear.

Reply:
okie dokie, just need to check what happens with the pact broker client. I know that for example --dry-run is implicitly set to false from the cli if not passed, should the env var overwrite that. I assumed it would, as it allows me to set an configurable env var in a CI action, that I can tweak and it can change the behaviour of the CLI, without me needing to make a code change (so I could allow a can-i-deploy step to have `PACT_BROKER_CAN_I_DEPLOY_DRY_RUN=true` and it take effect.

https://github.com/pact-foundation/pact_broker-client?tab=readme-ov-file#can-i-deploy